src/com/goide/inspections/GoPlaceholderChecker.java

```diff
@@ -0,0 +1,440 @@
+    private final String placeholder;
+    private final int startPos;
+    private final State state;
+    private final PrintVerb verb;
+    private final List<Integer> arguments;
+    private final String flags;
+
+    enum State {
+      OK,
+      MISSING_VERB_AT_END,
+      ARGUMENT_INDEX_NOT_NUMERIC
+    }
+
+    Placeholder(State state, int startPos, String placeholder, String flags, List<Integer> arguments, PrintVerb verb) {
+      this.placeholder = placeholder;
+      this.startPos = startPos;
+      this.verb = verb;
+      this.state = state;
+      this.arguments = arguments;
+      this.flags = flags;
+    }
+
+    public String getPlaceholder() {
+      return placeholder;
+    }
+
+    public int getPosition() {
+      return arguments.get(arguments.size() - 1);
+    }
+
+    public State getState() {
+      return state;
+    }
+
+    public PrintVerb getVerb() {
+      return verb;
+    }
+
+    public List<Integer> getArguments() {
+      return arguments;
+    }
+
+    public String getFlags() {
+      return flags;
+    }
+
+    public int getStartPos() {
+      return startPos;
+    }
+  }
+
+  @NotNull
+  public static List<Placeholder> parsePrintf(@NotNull String placeholderText) {
+    List<Placeholder> placeholders = new ArrayList<Placeholder>();
+    int argNum = 1;
+    int w;
+    for (int i = 0; i < placeholderText.length(); i += w) {
+      w = 1;
+      if (placeholderText.charAt(i) == '%') {
+        FormatState state = parsePrintfVerb(placeholderText.substring(i), i, argNum);
+        w = state.format.length();
+
+        // We are not interested in %% which prints %
+        if (state.state == Placeholder.State.OK && state.verb == PrintVerb.Percent) {
+          // Special magic case for allowing things like %*% to pass (which are sadly valid expressions)
+          if (state.format.length() == 2) continue;
+        }
+
+        placeholders.add(state.toPlaceholder());
+
+        // We only consider ok states as valid to increase the number of arguments. Should we?
+        if (state.state != Placeholder.State.OK) continue;
+
+        if (!state.indexed) {
+          int maxArgNum = Collections.max(state.argNums);
+          if (argNum < maxArgNum) {
+            argNum = maxArgNum;
+          }
+        }
+        else {
+          argNum = state.argNums.get(state.argNums.size() - 1);
+        }
+        argNum++;
+      }
+    }
+
+    return placeholders;
+  }
+
+  protected static int getPlaceholderPosition(@NotNull GoFunctionOrMethodDeclaration function) {
+    Integer position = FORMATTING_FUNCTIONS.get(StringUtil.toLowerCase(function.getName()));
+    if (position != null) {
+      String importPath = function.getContainingFile().getImportPath(false);
+      if ("fmt".equals(importPath) || "log".equals(importPath) || TESTING_PATH.equals(importPath)) {
+        return position;
+      }
+    }
+    return -1;
+  }
+
+  @Nullable
+  protected static String resolve(@NotNull GoExpression argument) {
```

Comment:
I was thinking this would also be useful in other use cases of this without the refactoring to move it here then. But sure I can move it.